File: java_src/limelight/Book.java
package limelight;

import javax.swing.*;
import javax.swing.filechooser.FileFilter;
import java.awt.event.ActionEvent;
import java.awt.event.ActionListener;
import java.io.File;

public class Book
{
	private JFrame frame;
  private File directory;

  public Book()
	{
		frame = new JFrame();

    JMenuBar menuBar = new JMenuBar();
    frame.setJMenuBar(menuBar);
    JMenu fileMenu = new JMenu("File");
    menuBar.add(fileMenu);
    JMenuItem open = new JMenuItem("Open");
    fileMenu.add(open);
    open.addActionListener(new ActionListener() {
      public void actionPerformed(ActionEvent event)
      {
        JFileChooser chooser = new JFileChooser();
        if(directory != null)
          chooser.setCurrentDirectory(directory);
        chooser.setFileFilter(new FileFilter() {
          public boolean accept(File file)
          {
            return file.getName().endsWith("llm");  
          }

          public String getDescription()
          {
            return "Limelight Markup File";
          }
        });
        int returnVal = chooser.showOpenDialog(frame);
        if(returnVal == JFileChooser.APPROVE_OPTION) {
          load2(chooser.getSelectedFile().getAbsolutePath());
          directory = chooser.getSelectedFile().getAbsoluteFile().getParentFile();
        }
      }
    });
    JMenuItem refresh = new JMenuItem("Refresh");
    refresh.addActionListener(new ActionListener() {
      public void actionPerformed(ActionEvent event)
      {
        reload();
      }
    });
    fileMenu.add(refresh);

  }

	public void open()
	{
		open(new DefaultPage());
	}

	public void open(Page page)
	{
		frame.setLayout(null);
		frame.setSize(900, 900);
		frame.setLocation(200, 25);
		frame.setDefaultCloseOperation(WindowConstants.EXIT_ON_CLOSE);
		loadPage(page);
		frame.setVisible(true);
		frame.repaint();
	}

	public void close()
	{
		frame.setVisible(false);
		frame.dispose();
	}

  public void load2(String filename)
  {
System.out.println("filename = " + filename);
    // to be overidden
  }

  public void reload()
  {
System.out.println("reload");
    // to be overidden
  }

  public void loadPage(Page page)
	{
		frame.getContentPane().removeAll();
		frame.add(page.getPanel());
		page.setBook(this);
		page.getPanel().setSize(frame.getSize());
	}

	public JFrame getFrame()
	{
		return frame;
	}
}

File: ruby_src/lib/limelight/book.rb
require 'limelight/java_util'
require 'limelight/menu_bar'
require 'limelight/loaders/file_scene_loader'

module Limelight
  
  class Book
    attr_accessor :directory, :default_scene, :styles
    attr_reader :frame, :current_scene, :producer
    
    def public_choose_file
      choose_file
    end
    
    def initialize(producer)
      @producer = producer
      @styles = {}
      @frame = javax.swing.JFrame.new
      @frame.setDefaultCloseOperation(javax.swing.WindowConstants::EXIT_ON_CLOSE)
      @frame.setLayout(nil)
      @frame.setSize(900, 900)
      @frame.setLocation(200, 25)
      
      menu_bar = MenuBar.build(self) do
        menu("File") do
          item("Open", :choose_file)
          item("Refresh", :reload)
        end
      end
      
      frame.setJMenuBar(menu_bar)
    end
    
    def open(scene)
      loadScene(scene)
      @frame.setVisible(true)
      @frame.repaint
      scene.visible = true
    end
  
    def close
      @frame.setVisible(false)
      @frame.dispose()
    end
    
    def loadScene(scene)
      @frame.content_pane.removeAll
      @frame.add(scene.panel)
      scene.book = self
      scene.panel.size = @frame.size
      @current_scene = scene
    end
    
    def load(scene_path)
      @producer.open_scene(scene_path)
    end
  
    def reload
      load(@current_scene.loader.scene_file)
    end

    private ###############################################
    
    def choose_file
      chooser = javax.swing.JFileChooser.new
      chooser.setCurrentDirectory(@directory) if @directory
      chooser.setFileFilter(ChooseFileFilter.new)
      returnVal = chooser.showOpenDialog(@frame);
      
      if(returnVal == javax.swing.JFileChooser::APPROVE_OPTION)
        load(chooser.getSelectedFile().getAbsolutePath());
        @directory = chooser.getSelectedFile().getAbsoluteFile().getParentFile();
      end
    end

  end
  
  class ChooseFileFilter < javax.swing.filechooser::FileFilter
    
    def accept(file)
      return file.name[-3..-1] == "llm"
    end
    
    def getDescription
      return "Limelight Markup File"
    end
  end
    
end

